staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,250 @@
+	// custom resource definition's OpenAPIv3 schema is recursively validated.
+	TotalCost *TotalCost
+}
+
+// Converter converts from JSON schema to a structural schema and a CEL declType, or returns an error if the conversion
+// fails.
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+// DefaultConverter converts using standard conversion utilities.
+var DefaultConverter = func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error) {
+	structural, err := structuralschema.NewStructural(schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	return structural, model.SchemaDeclType(structural, isRoot), nil
+}
+
+// RootCelContext constructs CelSchemaContext for the given root schema.
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CelSchemaContext {
+	rootCardinality := uint64(1)
+	r := CelSchemaContext{
+		MaxCardinality: &rootCardinality,
+		TotalCost:      &TotalCost{},
+		Converter:      converter,
+	}
+	r.initSchemaFields(schema, true)
+	return r
+}
+
+// ChildFieldContext constructs and returns a CelSchemaContext for fieldName.
+func (c *CelSchemaContext) ChildFieldContext(schema, propSchema *apiextensions.JSONSchemaProps, fieldName string) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Properties != nil {
+		if propSchema, ok := c.Schema.Properties[fieldName]; ok {
+			r.Schema = &propSchema
+		}
+		if declField, ok := c.DeclType.Fields[fieldName]; ok {
+			r.DeclType = declField.Type
+		}
+	} else {
+		r.initSchemaFields(propSchema, false)
+	}
+	return r
+}
+
+// ChildValuesContext constructs and returns a CelSchemaContext for the properties of an object if this CelSchemaContext is an object.
+func (c *CelSchemaContext) ChildValuesContext(schema *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.AdditionalProperties != nil {
+		if itemsSchema := c.Schema.AdditionalProperties; itemsSchema != nil {
+			r.Schema = itemsSchema.Structural
+		}
+		r.DeclType = c.DeclType.ElemType
+	} else {
+		r.initSchemaFields(schema.AdditionalProperties.Schema, false)
+	}
+	return r
+}
+
+// ChildItemsContext constructs and returns a CelSchemaContext for the items of an array if this CelSchemaContext is an array.
+func (c *CelSchemaContext) ChildItemsContext(schema, itemsSchema *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Items != nil {
+		if itemsSchema := c.Schema.Items; itemsSchema != nil {
+			r.Schema = itemsSchema
+		}
+		r.DeclType = c.DeclType.ElemType
+	} else {
+		r.initSchemaFields(itemsSchema, false)
+	}
+	return r
+}
+
+// ChildSchemaContext returns a CelSchemaContext for all child nodes that are not handled by ChildFieldContext,
+// ChildValuesContext or ChildItemsContext.
```

Comment:
if it's not a property, item, or additionalProperty, what other child nodes are there?

Reply:
not, allOf, oneOf and anyOf.  I've added to the godoc.